ladybug_geometry/polyskel.py

```diff
@@ -0,0 +1,694 @@
+				)
+
+			vertex.lav = lav
+			if lav.head == None:
+				lav.head = vertex
+				vertex.prev = vertex.next = vertex
+			else:
+				vertex.next = lav.head
+				vertex.prev = lav.head.prev
+				vertex.prev.next = vertex
+				lav.head.prev = vertex
+		return lav
+
+	@classmethod
+	def from_chain(cls, head, slav):
+		lav = cls(slav)
+		lav.head = head
+		for vertex in lav:
+			lav._len += 1
+			vertex.lav = lav
+		return lav
+
+	def invalidate(self, vertex):
+		assert vertex.lav is self, 'Tried to invalidate a vertex thats not mine'
+		log.debug('Invalidating %s', vertex)
+		vertex._valid = False
+		if self.head == vertex:
+			self.head = self.head.next
+		vertex.lav = None
+
+	def unify(self, vertex_a, vertex_b, point):
+		normed_b_bisector = vertex_b.bisector.v.normalize()
+		normed_a_bisector = vertex_a.bisector.v.normalize()
+		replacement =_LAVertex(
+			point,
+			vertex_a.edge_left,
+			vertex_b.edge_right,
+		 	(normed_b_bisector, normed_a_bisector)
+			)
+		replacement.lav = self
+
+		if self.head in [vertex_a, vertex_b]:
+			self.head = replacement
+
+		vertex_a.prev.next = replacement
+		vertex_b.next.prev = replacement
+		replacement.prev = vertex_a.prev
+		replacement.next = vertex_b.next
+
+		vertex_a.invalidate()
+		vertex_b.invalidate()
+
+		self._len -= 1
+		return replacement
+
+	def __str__(self):
+		return 'LAV {}'.format(id(self))
+
+	def __repr__(self):
+		return '{} = {}'.format(str(self), [vertex for vertex in self])
+
+	def __len__(self):
+		return self._len
+
+	def __iter__(self):
+		cur = self.head
+		while True:
+			yield cur
+			cur = cur.next
+			if cur == self.head:
+				raise StopIteration
+
+	def _show(self):
+		cur = self.head
+		while True:
+			print(cur.__repr__())
+			cur = cur.next
+			if cur == self.head:
+				break
+
+class _EventQueue:
+	def __init__(self):
+		self.__data = []
+
+	def put(self, item):
+		if item is not None:
+			heapq.heappush(self.__data, item)
+
+	def put_all(self, iterable):
+		for item in iterable:
+			heapq.heappush(self.__data, item)
+
+	def get(self):
+		return heapq.heappop(self.__data)
+
+	def empty(self):
+		return len(self.__data)==0
+
+	def peek(self):
+		return self.__data[0]
+
+	def show(self):
+		for item in self.__data:
+			print(item)
+
+
+def skeletonize(polygon, holes=None):
+	"""
+	Compute the straight skeleton of a polygon.
+
+	The polygon should be given as a list of vertices in counter-clockwise order.
+	Holes is a list of the contours of the holes, the vertices of which should be in
+	clockwise order.
+
+	Returns the straight skeleton as a list of "subtrees", which are in the form of
+	(source, height, sinks), where source is the highest points, height is its height,
+	and sinks are the point connected to the source.
+	"""
+
+	#HACK: But may be least bad way to add this functionality w/o contaminating
+	#rest of LB point geometry.
+	# Add polyskel-specific point2d magic method to Point2D used by heapq
+	Point2D.__lt__ = _polyskel_point2d_lt
```

Comment:
If you are going inject the code like this, let's just add the `__lt__` method to the `Point2D` class directly. I don't think it contaminates ladybug_geometry and we can just note in a docstring that the `<` operator is useful for sorting. We should add the `__gt__` operator as well so that the two are in agreement.

Reply:
The only issue is that choosing the x-coordinate for sorting is arbitrary, and may lead to undesired outcomes. So for example, if I have a set of points with the same x-coordinate, and different y-coordinates, and then I throw those points into a binary data tree, this operator will lead it to create an unbalanced tree, and no efficiency gained. So it seems like it'd be preferable to allow the user to modify the class on their own to choose the behaviour. 

The other thing is, I didn't have any luck tracking down an example of any current euclid.py implementation with this behaviour. It's showing up in the euclid.py file that Yongha has in his repo, but I strongly suspect he added that in there, and it just doesn't show up in the commits. So I don't think it's common behaviour.

Thoughts?

Reply:
There are three ways that I can see to resolve this but I'm confident that what is there now is not the best practice. When you say `Point2D.__lt__ = _polyskel_point2d_lt`, you are overwriting the `__lt__` method for all instances of the `Point2D` class and this will lead to a lot of confusion in terms of whether `<` is supported for the `Point2D` or not since it only gets added after a polyskeleton is computed. To put it another way, it's not good practice to overwrite hidden variables on a class when you are outside of that class. Granted, I know you do this a little in your code for single `_` variables, which have different values in each class instance. And I think this is fine because you know what you are doing and it helps a lot with your code's performance. But overwriting double `__` variables outside of the class is a whole other level of not-recommended practices. So I can see three solutions:

1. We add the `__lt__` to the `Point2D` class on the class itself and make it consistent with a `__gt__` method. I know that using the x-coordinate is arbitrary but I don't think most people have a pre-conceived notion about what it means for a point to be `<` or `>` to another and no other parts of the library have a need to use this operator for points right now. If we find in the future that there's a capability that needs a different definition of `<` or `>`, we can just add a different method under a new name.

2. Speaking of that solution, we could add a more explicit method to the `Point2D` class like `is_x_less_than` and try to use that.  It don't know exactly how the binary data tree works so I realize this might not be an easy-to-implement option. But I'm sure most future cases where we want a different definition of `<` will accommodate something like this.

3. We do this type of `<` evaluation in a function that you pass the point to. Again, I don't know if this is workable with your binary tree or if this would kill the performance of your code.

I am really ok with 1 and I can at least say it's a lot better than what is there right now.

Reply:
Okay let's do (1) then. Out of curiousity, is the overwriting of __lt__ just limited to all instances of Point2D created in this class, or somehow will overwrite all instances outside of the class as well? I did think of the ovewriting issue you raised, and figured it wouldn't be a problem since the input and output of the skeletonize will be a nested list of numbers, not geometry objects - and therefore this modification is limited to what's created here. 

Re: 2 and 3, I don't think either will work unfortunately. I also wanted to do something like this, but the particular data structure that uses this is being instantiated from a native python library (heapq) that only consumes input data (in this case the point2D) and then calls that method internally.

Reply:
FYI I also tried creating a custom PolyskelPoint2D class in the polyskel module, which inherited the Point2D class and added the __lt__ method. This didn't work, I believe because the LineSegments still used the original Point2D class and I figured it was getting too complicated a strategy to pursue.